internal/v2/controller/http/command.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package http
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	sdkCommon "github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/container"
+	"github.com/edgexfoundry/device-sdk-go/internal/v2/application"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	edgexErr "github.com/edgexfoundry/go-mod-core-contracts/errors"
+	"github.com/gorilla/mux"
+)
+
+const SDKPostEventReserved = "ds-postevent"
+const SDKReturnEventReserved = "ds-returnevent"
+
+func (c *V2HttpController) Command(writer http.ResponseWriter, request *http.Request) {
+	defer request.Body.Close()
+	vars := mux.Vars(request)
+	correlationID := request.Header.Get(sdkCommon.CorrelationHeader)
+
+	ds := container.DeviceServiceFrom(c.dic.Get)
+	err := checkServiceLocked(request, ds.AdminState)
+	if err != nil {
+		c.sendError(writer, request, edgexErr.KindServiceLocked, "service locked", err, sdkCommon.APIV2NameCommandRoute, correlationID)
+		return
+	}
+
+	var body string
+	var reserved url.Values
+	if request.Method == http.MethodPut {
+		// read request body for PUT command
+		body, err = readBodyAsString(request)
+		if err != nil {
+			c.sendError(writer, request, edgexErr.KindServerError, "failed to read request body", err, sdkCommon.APIV2NameCommandRoute, correlationID)
+			return
+		}
+	} else if request.Method == http.MethodGet {
+		// filter out the SDK reserved parameters and save the result for GET command
+		body, reserved = filterQueryParams(request.URL.RawQuery, c.lc)
+	}
+
+	isRead := request.Method == http.MethodGet
+	event, edgexErr := application.CommandHandler(isRead, vars, body, correlationID, c.dic)
+	if edgexErr != nil {
+		c.sendEdgexError(writer, request, edgexErr, sdkCommon.APIV2NameCommandRoute, correlationID)
+		return
+	}
+
+	// push to core and return event in http response based on SDK reserved query parameters
+	if ok, exist := reserved[SDKPostEventReserved]; exist && ok[0] == "yes" {
+		go sdkCommon.SendEvent(event, c.lc, container.CoredataEventClientFrom(c.dic.Get))
+	}
+	if ok, exist := reserved[SDKReturnEventReserved]; !exist || ok[0] == "yes" {
```

Comment:
This code should be in the App layer also.  Make it last thing CommandHandler does?